### PR TITLE
New Adapter: Adtargetme

### PR DESCRIPTION
**Pull request by grakholsky**
https://github.com/prebid/prebid.github.io/pull/3888

### CODE REVIEW

**Comment by SyntaxNode** [comment]
This PR is failing on the unit test:
```
--- FAIL: TestBidderUniquenessGatekeeping (0.00s)
    bidders_validate_test.go:62: 
        	Error Trace:	bidders_validate_test.go:62
        	Error:      	"9" is not less than or equal to "6"
        	Test:       	TestBidderUniquenessGatekeeping
```

This is because we require the first 6 characters of an bidder/adapter name to be unique unless both adapters refer to the same company. 

Is there a relationship between Adtarget and Adtargetme?

**Comment by grakholsky** [comment]
Hi.
We don’t have a relationship with Adtarget. 
[adtarget.me](https://adtarget.me/) is a different company.

What name should be correct in our case?

File: adapters/adserver_adtarget/adserver_adtarget.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+func getMediaTypeForImp(impId string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			} else if imp.Native != nil {
+				mediaType = openrtb_ext.BidTypeNative
+			}
+			return mediaType
```
Comment (AlexBVolcy) [inline adapters/adserver_adtarget/adserver_adtarget.go:186]: I might be confusing myself, but here it looks like you support `Video` and `Native` media types in addition to `Banner`. But in your docs PR and your `adserver_adtarget.yaml` file, it looks like you only support Banner. If you support video and native as well, could you update the docs PR and .yaml file? If not, could you explain what's being done here?

File: adapters/adserver_adtarget/adserver_adtarget.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		}
+	}
+	return mediaType
```
Comment (AlexBVolcy) [inline adapters/adserver_adtarget/adserver_adtarget.go:189]: I noticed that this return statement isn't covered by your current tests, could you add another JSON test to cover it?

File: adapters/adserver_adtarget/adserver_adtarget.go (hunk 1 above)
Comment (grakholsky) [reply adapters/adserver_adtarget/adserver_adtarget.go:186]: We want to start with a banner, but we added the logic beforehand for the video and native. We remove "video" and "native" support if this logic doesn't work.

**Comment by SyntaxNode** [comment]
The new name you've chosen `adserver_adtarget` does indeed start with 6 unique characters, but those characters are `adserv` which is a bit generic and we discourage symbols in the name. Perhaps you can remove the vowels in target to get `adtrgtme` or something similar rather prepending adserver.

**Review by AlexBVolcy — APPROVED** [review]
LGTM

**Comment by bsardo** [comment]
@grakholsky just a friendly reminder to give @SyntaxNode's comments a look when you get a chance.

File: adapters/adtrgtme/adtrgtme_test.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+	"testing"
+
+	"github.com/prebid/prebid-server/adapters/adapterstest"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+func TestJsonSamples(t *testing.T) {
+	bidder, buildErr := Builder(openrtb_ext.BidderAdtrgtme, config.Adapter{
+		Endpoint: "http://localhost:8077/ssp"})
```
[the remark this answers is not in the document]
Comment (grakholsky) [reply adapters/adtrgtme/adtrgtme_test.go]: We removed the port number from `adapters/adtrgtme/adtrgtmetest/**/*.json` and `adtrgtme_test.go`.

File: adapters/adtrgtme/adtrgtme.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	}
+
+	if response.StatusCode == http.StatusServiceUnavailable {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Something went wrong, please contact your Account Manager. Status Code: [ %d ] ", response.StatusCode),
+		}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return &errortypes.BadInput{
```
[the remark this answers is not in the document]
Comment (grakholsky) [reply adapters/adtrgtme/adtrgtme.go:125]: We added tests for handling other non-200 statuses:
- `adapters/adtrgtme/adtrgtmetest/supplemental/status-code-internal-server-error.json`
- `adapters/adtrgtme/adtrgtmetest/supplemental/status-code-temporary-redirect.json`

File: adapters/adtrgtme/adtrgtme.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if openRTBRequest.Device != nil {
+		if len(openRTBRequest.Device.UA) > 0 {
+			headers.Add("User-Agent", openRTBRequest.Device.UA)
+		}
+
+		if len(openRTBRequest.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", openRTBRequest.Device.IPv6)
```
[the remark this answers is not in the document]
Comment (grakholsky) [reply adapters/adtrgtme/adtrgtme.go:101]: We added:
- `adapters/adtrgtme/adtrgtmetest/supplemental/banner-app-headers-ipv6.json`
- `adapters/adtrgtme/adtrgtmetest/supplemental/banner-web-headers-ipv6.json`

**Review by AlexBVolcy — APPROVED** [review]

**Review by SyntaxNode — APPROVED** [review]

**Pull request merged**